File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/IpUtil.py
import socket
from typing import Optional

class IpUtil:
    @staticmethod
    def is_valid_ipv4(ip_address: str) -> bool:
        try:
            socket.getaddrinfo(ip_address, 0)
            return '.' in ip_address and ':' not in ip_address
        except socket.gaierror:
            return False

    @staticmethod
    def is_valid_ipv6(ip_address: str) -> bool:
        try:
            socket.getaddrinfo(ip_address, 0)
            return ':' in ip_address
        except socket.gaierror:
            return False

    @staticmethod
    def get_hostname(ip_address: str) -> Optional[str]:
        try:
            if ip_address == "0.0.0.0":
                return socket.getfqdn()
            
            addresses = socket.getaddrinfo(ip_address, 0, 0, 0, 0, socket.AI_CANONNAME)
            if not addresses:
                return None
            
            canonical_hostname = addresses[0][3]
            if canonical_hostname == ip_address:
                return None
            return canonical_hostname
        except socket.gaierror:
            return None

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_cpp/code_output/IpUtil.cpp
#pragma once

#include <string>
#include <optional>

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#pragma comment(lib, "ws2_32.lib")
#else
#include <sys/socket.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <unistd.h>
#endif

#ifndef NI_MAXHOST
#define NI_MAXHOST 1025
#endif

namespace org::example {

class IpUtil {
private:
    static void ensureWinsockInit() {
#ifdef _WIN32
        static struct Init {
            Init() {
                WSADATA wsaData;
                WSAStartup(MAKEWORD(2, 2), &wsaData);
            }
            ~Init() {
                WSACleanup();
            }
        } init;
#endif
    }

public:
    static bool isValidIpv4(const std::string& ipAddress) {
        ensureWinsockInit();

        struct addrinfo hints{}, *result = nullptr;
        hints.ai_family = AF_UNSPEC;

        int status = getaddrinfo(ipAddress.c_str(), nullptr, &hints, &result);
        if (status != 0) {
            return false;
        }
        freeaddrinfo(result);

        return ipAddress.find('.') != std::string::npos
            && ipAddress.find(':') == std::string::npos;
    }

    static bool isValidIpv6(const std::string& ipAddress) {
        ensureWinsockInit();

        struct addrinfo hints{}, *result = nullptr;
        hints.ai_family = AF_UNSPEC;

        int status = getaddrinfo(ipAddress.c_str(), nullptr, &hints, &result);
        if (status != 0) {
            return false;
        }
        freeaddrinfo(result);

        return ipAddress.find(':') != std::string::npos;
    }

    static std::optional<std::string> getHostname(const std::string& ipAddress) {
        ensureWinsockInit();

        if (ipAddress == "0.0.0.0") {
            char hostname[256];
            if (gethostname(hostname, sizeof(hostname)) == 0) {
                return std::string(hostname);
            }
            return std::nullopt;
        }

        struct addrinfo hints{}, *result = nullptr;
        hints.ai_family = AF_UNSPEC;

        int status = getaddrinfo(ipAddress.c_str(), nullptr, &hints, &result);
        if (status != 0) {
            return std::nullopt;
        }

        char host[NI_MAXHOST];
        status = getnameinfo(result->ai_addr, result->ai_addrlen,
                             host, sizeof(host), nullptr, 0, 0);
        freeaddrinfo(result);

        if (status != 0) {
            return std::nullopt;
        }

        std::string hostname(host);
        if (hostname == ipAddress) {
            return std::nullopt;
        }
        return hostname;
    }
};

} // namespace org::example